File: Java Galaga/enemyGrid.java
import java.util.Random;

public class enemyGrid {
	
	private SoundManager SOUND_MANAGER;
	
	// positions of grid locations
	private double[][] enemiesX;
	private double[][] enemiesY;
	
	private int leftBound;
	private int rightBound;
	private int distanceBetweenPoints;
	private int breatheCurrentStep = 1;
	
	// how far the grid will "breathe"
	private int breatheSpread = 35;
	
	private double breatheDelta;
	private double breathYDelta;
	
	// frames it will take to complete one down breathing motion or one up breathing motion
	private double breatheMaxSteps = 180;
	
	//how far the grid will breathe in they direction compared to the x values
	private double ySpreadRatio = 1.6;
	
	private boolean movingRight = true;
	private boolean shouldStartBreathing = false;
	private boolean breatheForRestOfRound = false;
	private boolean breathingDown = true;
	private Random random;
	
public enemyGrid(int distanceBetweenPoints, int leftBound, int rightBound, SoundManager SOUND_MANAGER)
{
	random = new Random();
	this.SOUND_MANAGER = SOUND_MANAGER;
	this.leftBound = leftBound;
	this.rightBound = rightBound;
	this.distanceBetweenPoints = distanceBetweenPoints;

	enemiesX = new double[5][10];
	enemiesY = new double[5][10];
	

	 
	 // How much Y will move each breatheStep, used for calculating future positions
	 breathYDelta = breatheSpread / ySpreadRatio / breatheMaxSteps;
}


public void update()
{
	
	// when we actually start breathing, which is when all enemies are on the grid and the grid's middle is in the middle of the screen
	if(shouldStartBreathing && enemiesX[0][5]==((leftBound+rightBound)/2)+25)
		breatheForRestOfRound = true;
	
	// breathing
	if(breatheForRestOfRound)
	{
			if(breathingDown)
			{
				 for (int row = 0; row < enemiesX.length; row++) {
					    for (int col = 0; col < enemiesX[row].length; col++) {
					    	if(col>4)
					    	{
					    		breatheDelta = (breatheSpread * Math.abs(col-5)) / breatheMaxSteps;
					    		enemiesX[row][col] += breatheDelta ;
					    	}
					    	else
					    	{
					    		breatheDelta = (breatheSpread * Math.abs(col-5))/ breatheMaxSteps;
					    		enemiesX[row][col] -= breatheDelta ;
					    	}
					    }
					 }
				 
			 for (int row = 0; row < enemiesY.length; row++) {
				    for (int col = 0; col < enemiesY[row].length; col++) {
				    	breatheDelta = (breatheSpread/ySpreadRatio*(row +1))/breatheMaxSteps;
				    	enemiesY[row][col] += breatheDelta;
				    }
				 }
			}
			
			else
			{
				for (int row = 0; row < enemiesX.length; row++) {
				    for (int col = 0; col < enemiesX[row].length; col++) {
					    	if(col>4)
					    	{
					    		breatheDelta = (breatheSpread * Math.abs(col-5)) / breatheMaxSteps;
					    		enemiesX[row][col] -= breatheDelta;
					    	}
					    	else
					    	{
					    		breatheDelta = (breatheSpread * Math.abs(col-5))/ breatheMaxSteps;
					    		enemiesX[row][col] += breatheDelta;
					    	}
				    	}
				    }
				
				for (int row = 0; row < enemiesY.length; row++) {
					for (int col = 0; col < enemiesY[row].length; col++) {
							breatheDelta =(breatheSpread/ySpreadRatio*(row +1))/breatheMaxSteps;
							enemiesY[row][col]-= breatheDelta; 
						} 
					}
				
			}
			
			if(breathingDown)
			breatheCurrentStep++;
			else breatheCurrentStep--;
			
			if(breatheCurrentStep ==0)
			{
				breathingDown = true;
				SOUND_MANAGER.breathingDown.play();
			}
			if(breatheCurrentStep == breatheMaxSteps)
				breathingDown = false;

		
		
	}
	else 
	{
		for (int row = 0; row < enemiesX.length; row++) {
		    for (int col = 0; col < enemiesX[row].length; col++) {
		    	
		    	if(movingRight)
		    	{
		    		enemiesX[row][col] +=1;
		    	}
		    	else
		    		enemiesX[row][col] -=1;
		    	
		    	
		    }
		 }
		if(enemiesX[0][0]<=leftBound)
		    movingRight = true;
		
		if(enemiesX[0][9]>= rightBound)
			movingRight = false;
	}
}

public int getXCord(int row, int col)
{
	if(row ==-1 || col ==-1)
		return -1;
	return (int) enemiesX[row][col];
}
public int getYCord(int row, int col)
{
	if(row ==-1 || col ==-1)
		return -1;
	return (int) enemiesY[row][col];
}
public void setToBreathe()
{
	shouldStartBreathing = true;
}

// gives the x cord of a grid location in a number of frames in the future
public int calcXCordInFrames(int row, int col, int numOfFrames)
{
	// handles x position prediction when the grid is just moving between the bounds(not breathing)
	if(!breatheForRestOfRound)
	{
		if(movingRight)
		{
			if(enemiesX[0][9]+numOfFrames >= rightBound)
				return (int) (enemiesX[row][col] - (numOfFrames - (rightBound-enemiesX[0][9]))+rightBound-enemiesX[0][9]);
				else
					return (int) (enemiesX[row][col]+numOfFrames);
		}
		else
		{
			if(enemiesX[1][0]-numOfFrames <= leftBound)
			{
				
				return (int) (enemiesX[row][col] + (numOfFrames - (enemiesX[1][0]-leftBound)) -(enemiesX[1][0]-leftBound));
			}
				else
					return (int) (enemiesX[row][col]-numOfFrames);
		}
	}
	
	// handles x position prediction when the grid is breathing
	else
	{
		
		// grid currently breathing down
		if(breathingDown)
		{
			// checks if the grid doesn't switch breathing directions in the amount of frames specified
			if(breatheMaxSteps-breatheCurrentStep>numOfFrames)
			{
				// since were not expecting a shift, use the original formula of breathing to predict where we will be in numOfFrames
				if(col<5)
				return (int) (enemiesX[row][col] - (numOfFrames)*(breatheSpread * Math.abs(col-5)/breatheMaxSteps));
				else 
					return (int) (enemiesX[row][col] + (numOfFrames)*(breatheSpread * Math.abs(col-5)/breatheMaxSteps));
			}
			// grid will shift directions during the amount of frames specified
			// handled by going in the direction expected for(breatheMaxSteps-breatheCurrentStep)(how many steps left we have till the grid shifts breathing directions)
			// then going in the opposite direction by (numOfFrames-(breatheMaxSteps-breatheCurrentStep)) which is how many frames we have left after doing the first breathing direction
			else
			{
				if(col<5)
				return (int) (enemiesX[row][col] +(-(breatheMaxSteps-breatheCurrentStep)*(breatheSpread * Math.abs(col-5)/breatheMaxSteps) 
						+(numOfFrames-(breatheMaxSteps-breatheCurrentStep))*(breatheSpread * Math.abs(col-5)/breatheMaxSteps)));
				else 
					return
							(int) (enemiesX[row][col] +((breatheMaxSteps-breatheCurrentStep)*(breatheSpread * Math.abs(col-5)/breatheMaxSteps) 
									-(numOfFrames-(breatheMaxSteps-breatheCurrentStep))*(breatheSpread * Math.abs(col-5)/breatheMaxSteps)));
			}
		}
		// grid currently breathing up
		else
		{
			// checks if the grid doesn't switch breathing directions in the amount of frames specified
			if(breatheCurrentStep>=numOfFrames)
			{
				// since were not expecting a shift, use the original formula of breathing to predict where we will be in numOfFrames
				if(col<5)
				return (int) (enemiesX[row][col] + (numOfFrames)*(breatheSpread * Math.abs(col-5)/breatheMaxSteps));
				else 
					return (int) (enemiesX[row][col] - (numOfFrames)*(breatheSpread * Math.abs(col-5)/breatheMaxSteps));
			}
			
			// grid will shift directions during the amount of frames specified 
			// handled by going in the direction expected for breatheCurrentStep amount then going in (numOfFrames-breatheCurrentStep) since we expect the numOfFrames to be larger
			else
			{
				if(col<4)
				return (int) (enemiesX[row][col] + ((breatheCurrentStep)*(breatheSpread * Math.abs(col-5)/breatheMaxSteps) 
						-(numOfFrames-breatheCurrentStep)*(breatheSpread * Math.abs(col-5)/breatheMaxSteps)));
				else 	
					return (int) (enemiesX[row][col] + (-(breatheCurrentStep)*(breatheSpread * Math.abs(col-5)/breatheMaxSteps) 
									+(numOfFrames-breatheCurrentStep)* (breatheSpread * Math.abs(col-5)/breatheMaxSteps)));
			}
		}
	}
}


public int calcYCordInFrames(int row, int col, int numOfFrames)
{
	if(isBreathing())
	{
		if(breathingDown)
		{
			if(breatheMaxSteps-breatheCurrentStep>=numOfFrames)
			{
				return (int) (enemiesY[row][col] + (numOfFrames * (breathYDelta*(row +1))));
			}
			else
			{
				return (int) (enemiesY[row][col] + (((breatheMaxSteps-breatheCurrentStep)*(breathYDelta*(row +1)) - 
						(numOfFrames-(breatheMaxSteps-breatheCurrentStep)) * (breathYDelta*(row + 1)))));
			}
		}
		else
		{
			if(breatheCurrentStep>=numOfFrames)
			{
				return (int) (enemiesY[row][col] - (numOfFrames*(breathYDelta*(row +1))));
			}
			else 
				return (int) (enemiesY[row][col] + (-(breathYDelta*(row +1) * breatheCurrentStep) + 
						(numOfFrames-breatheCurrentStep)* (breathYDelta*(row +1))));
		}
	}
	else
		//when the grid isn't breathing
		return getYCord(row,col);
}
public void reset()
{
	movingRight = random.nextBoolean();
	shouldStartBreathing = false;
	breatheForRestOfRound = false;
	breathingDown = true;
	breatheCurrentStep = 1;
	int seed = random.nextInt(400);
	 for (int row = 0; row < enemiesX.length; row++) {
		    for (int col = 0; col < enemiesX[row].length; col++) {
		    	enemiesX[row][col] = seed+100+(distanceBetweenPoints*col);
		    }
		 }
	 
	 for (int row = 0; row < enemiesY.length; row++) {
		    for (int col = 0; col < enemiesY[row].length; col++) {
		    	enemiesY[row][col] = 50+(50*row);
		    }
		 }
}

public boolean isBreathing()
{
	return breatheForRestOfRound;
}

public boolean isSetToBreathe()
{
	return shouldStartBreathing;
}
public boolean isBreathingDown()
{
	return breathingDown;
}

}

File: python/DQNHalf/galaga-gym/galaga_gym/envs/EnemyGrid.py
import numpy as np
import random

class EnemyGrid:
  def __init__(self, distanceBetweenPoints, leftBound, rightBound):
    self.enemiesX = np.empty((5, 10))
    self.enemiesY = np.empty((5, 10))

    self.leftBound = leftBound
    self.rightBound = rightBound

    self.distanceBetweenPoints = distanceBetweenPoints
    self.breatheCurrentStep = 1

    self.breatheSpread = 35

    self.breatheDelta = 0.0
    
    self.breatheMaxSteps = 180

    self.ySpreadRatio = 1.6

    self.movingRight = True
    self.shouldStartBreathing = False
    self.breatheForRestOfRound = False
    self.breathingDown = True

    self.breatheYDelta = self.breatheSpread / self.ySpreadRatio / self.breatheMaxSteps
  
  def update(self):
    if self.shouldStartBreathing and self.enemiesX[0][5] == int((self.leftBound + self.rightBound)/2) + 25:
      self.breatheForRestOfRound = True
    
    if self.breatheForRestOfRound:
      if self.breathingDown:
        for row in range(self.enemiesX.shape[0]):
          for col in range(self.enemiesX.shape[1]):
            self.breatheDelta = (self.breatheSpread * abs(col - 5)) / self.breatheMaxSteps
            if col > 4:
              self.enemiesX[row][col] += self.breatheDelta
            else:
              self.enemiesX[row][col] -= self.breatheDelta
        
        for row in range(self.enemiesY.shape[0]):
          for col in range(self.enemiesY.shape[1]):
            self.breatheDelta = (self.breatheSpread / self.ySpreadRatio * (row+1))/self.breatheMaxSteps
            self.enemiesY[row][col] += self.breatheDelta
        
      else:
        for row in range(self.enemiesX.shape[0]):
          for col in range(self.enemiesX.shape[1]):
            self.breatheDelta = (self.breatheSpread * abs(col-5)) / self.breatheMaxSteps
            if col > 4: self.enemiesX[row][col] -= self.breatheDelta
            else:       self.enemiesX[row][col] += self.breatheDelta
        
        for row in range(self.enemiesY.shape[0]):
          for col in range(self.enemiesY.shape[1]):
            self.breatheDelta = (self.breatheSpread / self.ySpreadRatio * (row+1))/self.breatheMaxSteps
            self.enemiesY[row][col] -= self.breatheDelta

      self.breatheCurrentStep += 1 if self.breathingDown else -1

      if self.breatheCurrentStep == 0:
        self.breathingDown = True
      elif self.breatheCurrentStep == self.breatheMaxSteps:
        self.breathingDown = False
    
    else:
      for row in range(self.enemiesX.shape[0]):
        for col in range(self.enemiesX.shape[1]):
          if self.movingRight: self.enemiesX[row][col] += 1
          else:                self.enemiesX[row][col] -= 1
      
      if self.enemiesX[0][0] <= self.leftBound:  self.movingRight = True
      if self.enemiesX[0][9] >= self.rightBound: self.movingRight = False
  
  def getXCord(self, row, col):
    return int(self.enemiesX[row][col])
  
  def getYCord(self, row, col):
    return int(self.enemiesY[row][col])
  
  def setToBreathe(self):
    self.shouldStartBreathing = True
  
  def calcXCordInFrames(self, row, col, numOfFrames):
    if not self.breatheForRestOfRound:
      if self.movingRight:
        if self.enemiesX[0][9] + numOfFrames >= self.rightBound:
          return int(self.enemiesX[row][col] - (numOfFrames - (self.rightBound - self.enemiesX[0][9])) + self.rightBound - self.enemiesX[0][9])
        else:
          return int(self.enemiesX[row][col] + numOfFrames)
      
      else:
        if self.enemiesX[1][0] - numOfFrames <= self.leftBound:
          return int(self.enemiesX[row][col] + (numOfFrames - (self.enemiesX[1][0]-self.leftBound)) - (self.enemiesX[1][0] - self.leftBound))
        else:
          return int(self.enemiesX[row][col] - numOfFrames)
    
    else:
      if self.breathingDown:
        if self.breatheMaxSteps - self.breatheCurrentStep > numOfFrames:
          if col < 5: return int(self.enemiesX[row][col] - numOfFrames * (self.breatheSpread * abs(col - 5) / self.breatheMaxSteps))
          else:       return int(self.enemiesX[row][col] + numOfFrames * (self.breatheSpread * abs(col - 5) / self.breatheMaxSteps))
        
        else:
          if col < 5:
            return int(self.enemiesX[row][col] + (-(self.breatheMaxSteps - self.breatheCurrentStep)*(self.breatheSpread * abs(col - 5) / self.breatheMaxSteps) +
                (numOfFrames - (self.breatheMaxSteps - self.breatheCurrentStep)) * (self.breatheSpread * abs(col - 5) / self.breatheMaxSteps)))
          else:
            return int(self.enemiesX[row][col] + ((self.breatheMaxSteps-self.breatheCurrentStep)*(self.breatheSpread * abs(col - 5)/self.breatheMaxSteps) -
                (numOfFrames - (self.breatheMaxSteps - self.breatheCurrentStep)) * (self.breatheSpread * abs(col - 5) / self.breatheMaxSteps)))
      else:
        if self.breatheCurrentStep >= numOfFrames:
          if col < 5:
            return int(self.enemiesX[row][col] + numOfFrames*(self.breatheSpread * abs(col - 5)/self.breatheMaxSteps))
          else:
            return int(self.enemiesX[row][col] - numOfFrames*(self.breatheSpread * abs(col - 5)/self.breatheMaxSteps))
        
        else:
          if col < 4:
            return int(self.enemiesX[row][col] + ((self.breatheCurrentStep) * (self.breatheSpread * abs(col - 5) / self.breatheMaxSteps)
                -(numOfFrames - self.breatheCurrentStep) * (self.breatheSpread * abs(col - 5)/self.breatheMaxSteps)))
          else:
            return int(self.enemiesX[row][col] + (-(self.breatheCurrentStep)*(self.breatheSpread * abs(col - 5) / self.breatheMaxSteps)
                +(numOfFrames - self.breatheCurrentStep) * (self.breatheSpread * abs(col - 5)/self.breatheMaxSteps)))
  
  def calcYCordInFrames(self, row, col, numOfFrames):
    if self.isBreathing():
      if self.breathingDown:
        if self.breatheMaxSteps - self.breatheCurrentStep >= numOfFrames:
          return int(self.enemiesY[row][col] + (numOfFrames * (self.breatheYDelta*(row + 1))))
        else:
          return int(self.enemiesY[row][col] + (((self.breatheMaxSteps - self.breatheCurrentStep) * (self.breatheYDelta*(row+1)) -
              (numOfFrames - (self.breatheMaxSteps - self.breatheCurrentStep)) * (self.breatheYDelta*(row + 1)))))
      else:
        if self.breatheCurrentStep >= numOfFrames:
          return int(self.enemiesY[row][col] - (numOfFrames * (self.breatheYDelta*(row+1))))
        else:
          return int(self.enemiesY[row][col] + (-(self.breatheYDelta*(row+1) * self.breatheCurrentStep) +
              (numOfFrames - self.breatheCurrentStep) * (self.breatheYDelta*(row+1))))
    
    else:
      return self.getYCord(row, col)
  
  def reset(self):
    self.movingRight = bool(random.getrandbits(1))
    self.shouldStartBreathing = False
    self.breatheForRestOfRound = False
    self.breathingDown = True
    self.breatheCurrentStep = 1
    seed = random.randint(0, 400)

    for row in range(self.enemiesX.shape[0]):
      for col in range(self.enemiesX.shape[1]):
        self.enemiesX[row][col] = seed + 100 + (self.distanceBetweenPoints * col)
    
    for row in range(self.enemiesY.shape[0]):
      for col in range(self.enemiesY.shape[1]):
        self.enemiesY[row][col] = 50 + (50 * row)
  
  def isBreathing(self):
    return self.breatheForRestOfRound
  
  def isSetToBreathe(self):
    return self.shouldStartBreathing
  
  def isBreathingDown(self):
    return self.breathingDown